File: apps/mac-menubar/Sources/DaylightMenuBar/MoonPhase.swift
// Moon phase calculation for menu bar status rendering.
// Exports: MoonPhaseCalculator
// Deps: Foundation Date

import Foundation

final class MoonPhaseCalculator {
    private let j2000 = Date(timeIntervalSince1970: 946_728_000) // 2000-01-01 12:00 UTC

    /// Illuminated fraction 0…1 for the given date.
    func illuminatedFraction(for date: Date) -> Double {
        (1 - cos(2 * .pi * phaseValue(for: date))) / 2
    }

    /// True while the moon is waxing (new → full).
    func isWaxing(for date: Date) -> Bool {
        phaseValue(for: date) < 0.5
    }

    /// Phase 0…1 (0 = new, 0.5 = full) from the true sun–moon elongation,
    /// using the truncated series from Meeus, "Astronomical Algorithms"
    /// (~0.3° ≈ 35 minutes of phase). Mirrors getMoonPhase in core-calendar;
    /// the previous mean-phase approximation drifted up to ±1 day.
    private func phaseValue(for date: Date) -> Double {
        let d = date.timeIntervalSince(j2000) / 86_400
        let rad = Double.pi / 180
        let sunAnomaly = (357.529 + 0.98560028 * d) * rad
        let moonAnomaly = (134.963 + 13.064993 * d) * rad
        let meanElongation = (297.8502 + 12.190749 * d) * rad
        let latitudeArgument = (93.272 + 13.22935 * d) * rad
        let moonLongitude = 218.316 + 13.176396 * d
            + 6.289 * sin(moonAnomaly)
            + 1.274 * sin(2 * meanElongation - moonAnomaly)
            + 0.658 * sin(2 * meanElongation)
            + 0.214 * sin(2 * moonAnomaly)
            - 0.186 * sin(sunAnomaly)
            - 0.114 * sin(2 * latitudeArgument)
        let sunLongitude = 280.459 + 0.98564736 * d
            + 1.915 * sin(sunAnomaly)
            + 0.02 * sin(2 * sunAnomaly)
        return normalizedPhase((moonLongitude - sunLongitude) / 360)
    }

    private func normalizedPhase(_ value: Double) -> Double {
        let phase = value - floor(value)
        return phase < 0 ? phase + 1 : phase
    }
}

File: packages/core-calendar/src/moon-phase.ts
/**
 * Moon phase calculation for app icons and date detail metadata.
 * Exports: getMoonPhase
 * Deps: domain calendar types, date-key helpers
 */

import type { LocalDateKey, MoonPhase, MoonPhaseName } from "@daylight/domain";
import { toUtcDate } from "./date-key";

const J2000_UTC_MS = Date.UTC(2000, 0, 1, 12);
const DEGREES_TO_RADIANS = Math.PI / 180;
const SYNODIC_MONTH_DAYS = 29.530588853;

export function getMoonAgeDays(date: LocalDateKey): number {
  return SYNODIC_MONTH_DAYS * getMoonPhase(date).phase;
}

export function getMoonPhase(date: LocalDateKey): MoonPhase {
  const daysSinceJ2000 = (toUtcDate(date).getTime() - J2000_UTC_MS) / 86_400_000;
  const normalizedPhase = normalizePhase(sunMoonElongationDegrees(daysSinceJ2000) / 360);
  const illumination = (1 - Math.cos(normalizedPhase * 2 * Math.PI)) / 2;
  return {
    name: classifyMoonPhase(normalizedPhase),
    illumination,
    phase: normalizedPhase,
  };
}

/**
 * Geocentric elongation of the Moon from the Sun in ecliptic longitude, using
 * the truncated series from Meeus, "Astronomical Algorithms" (low-precision
 * lunar longitude, ~0.3° ≈ 35 minutes of phase). The previous mean-phase
 * approximation drifted up to ±1 day from the true phase.
 */
function sunMoonElongationDegrees(daysSinceJ2000: number): number {
  const d = daysSinceJ2000;
  const sunAnomaly = (357.529 + 0.98560028 * d) * DEGREES_TO_RADIANS;
  const moonAnomaly = (134.963 + 13.064993 * d) * DEGREES_TO_RADIANS;
  const meanElongation = (297.8502 + 12.190749 * d) * DEGREES_TO_RADIANS;
  const latitudeArgument = (93.272 + 13.22935 * d) * DEGREES_TO_RADIANS;
  const moonLongitude =
    218.316 +
    13.176396 * d +
    6.289 * Math.sin(moonAnomaly) +
    1.274 * Math.sin(2 * meanElongation - moonAnomaly) +
    0.658 * Math.sin(2 * meanElongation) +
    0.214 * Math.sin(2 * moonAnomaly) -
    0.186 * Math.sin(sunAnomaly) -
    0.114 * Math.sin(2 * latitudeArgument);
  const sunLongitude =
    280.459 + 0.98564736 * d + 1.915 * Math.sin(sunAnomaly) + 0.02 * Math.sin(2 * sunAnomaly);
  return moonLongitude - sunLongitude;
}

function normalizePhase(value: number): number {
  const phase = value - Math.floor(value);
  return phase < 0 ? phase + 1 : phase;
}

function classifyMoonPhase(phase: number): MoonPhaseName {
  if (phase < 0.03 || phase >= 0.97) return "new";
  if (phase < 0.22) return "waxing-crescent";
  if (phase < 0.28) return "first-quarter";
  if (phase < 0.47) return "waxing-gibbous";
  if (phase < 0.53) return "full";
  if (phase < 0.72) return "waning-gibbous";
  if (phase < 0.78) return "last-quarter";
  return "waning-crescent";
}
